File: SortingAlgorithm/MergeSort.c
#include "stdio.h"

#define N 15

// void merge_sort(int a[], int b[], int start, int end){
//     if (start >= end)
//         return;

//     int     len = end - start;
//     int     mid = (len >> 1) + start;
//     int     start1 = start;
//     int     end1 = mid;
//     int     start2 = mid + 1;
//     int     end2 = end;
//     merge_sort(a, b, start1, end1);
//     merge_sort(a, b, start2, end2);

//     int     k = start;
//     while(start1 <= end1 && start2 <= end2)
//         b[k++] = a[start1] < a[start2] ? a[start1++] : a[start2++];
//     while(start1 <= end1)
//         b[k++] = a[start1++];
//     while(start2 <= end2)
//         b[k++] = a[start2++];

//     for (k = start; k <= end; k++)
//         a[k] = b[k];
// }

// void merge_sort(int a[], int b[], int start, int end){
//     if (start < end){
//         int mid = (start + end) / 2;
//         merge_sort(a, b, start, mid);
//         merge_sort(a, b, mid+1, end);

//         int i = start;
//         int j = mid + 1;
//         int m = mid;
//         int n = end;

//         int k = start;

//         while(i <= m && j <= n)
//             b[k++] = a[i] < a[j] ? a[i++] : a[j++];
//         while(i <= m)
//             b[k++] = a[i++];
//         while(j <= n)
//             b[k++] = a[j++];

//         for (i = start; i <= end; ++i)
//             a[i] = b[i];

//     }
// }

// void mergeSort(int a[]){
//     int b[N];
//     merge_sort(a, b, 0, N-1);
// }

void mergeSort(int a[]){
    int     b[N];
    int     i, j, m, n, x, y, t;

    for (i = 1; i < N; i *= 2)
    {
        for (j = 0; j < N; j += 2*i)
        {
            x = n = j + i < N ? j + i : N;
            y = j + i*2 < N ? j + i*2 : N;

            t = m = j;

            while(m < n && x < y)
                b[t++] = a[m] < a[x] ? a[m++] : a[x++];
            while(m < n)
                b[t++] = a[m++];
            while(x < y)
                b[t++] = a[x++];
        }

        for (j = 0; j < N; ++j)
        {
            a[j] = b[j];
        }
    }

}

int main(int argc, char const *argv[])
{
    int         a[N] = {10, 1, 23, -5, 0, 78, 11, 104, 65, -1, 12, 23, 36, 3, 53};

    printf("Origin : ");
    for (int i = 0; i < N; ++i)
    {
        printf("%4d", a[i]);
    }

    mergeSort(a);

    printf("\nSorted : ");
    for (int i = 0; i < N; ++i)
    {
        printf("%4d", a[i]);
    }

    return 0;
}

File: SortingAlgorithm/MergeSort.py
# coding=utf-8


# def merge_sort(a, b, start, end):
#     if start >= end:
#         return

#     mid = (start + end) / 2
#     merge_sort(a, b, start, mid)
#     merge_sort(a, b, mid+1, end)

#     k = m = start
#     n = mid
#     x = mid + 1
#     y = end

#     while (m <= n and x <= y):
#         if a[m] < a[x]:
#             b[k] = a[m]
#             m += 1
#         else:
#             b[k] = a[x]
#             x += 1
#         k += 1

#     while (m <= n):
#         b[k] = a[m]
#         k += 1
#         m += 1

#     while (x <= y):
#         b[k] = a[x]
#         k += 1
#         x += 1

#     for i in xrange(start, end+1):
#         a[i] = b[i]


# def mergeSort(array):
#     b = [0 for i in xrange(len(array))]
#     merge_sort(array, b, 0, len(array)-1)


def mergeSort(array):
    b = [0 for i in xrange(len(array))]

    i = 1
    while 1:
        if i >= len(array):
            break

        for j in xrange(0, len(b), 2*i):

            k = m = j
            n = x = j + i if j + i < len(b) else len(b)
            y = j + 2*i if j + 2*i < len(b) else len(b)

            while (m < n and x < y):
                if array[m] < array[x]:
                    b[k] = array[m]
                    m += 1
                else:
                    b[k] = array[x]
                    x += 1
                k += 1

            while (m < n):
                b[k] = array[m]
                k += 1
                m += 1

            while (x < y):
                b[k] = array[x]
                k += 1
                x += 1

        for j in xrange(len(b)):
            array[j] = b[j]

        i *= 2


if __name__ == '__main__':
    a = [10, 1, 23, -5, 0, 78, 11, 104, 65, -1, 12, 23, 36, 3, 53]
    print a

    mergeSort(a)
    print a
